### PR TITLE
PATE

### PR DESCRIPTION
**Pull request by kamathhrishi**
In reference to #1963 

This ports the numpy version of PATE to Torch.  

@iamtrask

### CODE REVIEW

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 1]
```diff
@@ -64,12 +65,19 @@ def compute_q_noisy_max(counts, noise_eps):
     # tight-bounds-on-probability-of-sum-of-laplace-random-variables
 
     winner = np.argmax(counts)
+    # print("winner")
```
Comment (mari-linhares) [inline syft/frameworks/torch/differential_privacy/pate.py]: Remove print statements

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 2]
```diff
@@ -64,12 +65,19 @@ def compute_q_noisy_max(counts, noise_eps):
...
+    # print(winner)
     counts_normalized = noise_eps * (counts - counts[winner])
+
+    # print("counts_normalized")
+    # print(counts_normalized)
```
Comment (mari-linhares) [inline syft/frameworks/torch/differential_privacy/pate.py]: same

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 3]
```diff
@@ -64,12 +65,19 @@ def compute_q_noisy_max(counts, noise_eps):
...
     counts_rest = np.array([counts_normalized[i] for i in range(len(counts)) if i != winner])
     q = 0.0
     for c in counts_rest:
         gap = -c
+
```
Comment (mari-linhares) [inline syft/frameworks/torch/differential_privacy/pate.py]: are these extra empty lines really needed? :thinking:

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 4]
```diff
@@ -266,3 +278,245 @@ def perform_analysis(teacher_preds, indices, noise_eps, delta=1e-5, moments=8, b
     # print("Data independent bound = " + str(min(data_ind_eps_list)) + ".")
 
     return min(eps_list_nm), min(data_ind_eps_list)
+
+
+def tensors_to_literals(tensor_list):
+
```
Comment (mari-linhares) [inline syft/frameworks/torch/differential_privacy/pate.py]: Remove empty line.

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 5]
```diff
@@ -266,3 +278,245 @@ def perform_analysis(teacher_preds, indices, noise_eps, delta=1e-5, moments=8, b
...
+    """Converts list of torch tensors to list of integers/floats. Fix for not having the functionality which converts list of tensors to tensors
+    
+       args:
```
Comment (mari-linhares) [inline syft/frameworks/torch/differential_privacy/pate.py]: args -> Args

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 6]
```diff
@@ -266,3 +278,245 @@ def perform_analysis(teacher_preds, indices, noise_eps, delta=1e-5, moments=8, b
     # print("Data independent bound = " + str(min(data_ind_eps_list)) + ".")
 
     return min(eps_list_nm), min(data_ind_eps_list)
+
+
+def tensors_to_literals(tensor_list):
```
Comment (mari-linhares) [inline syft/frameworks/torch/differential_privacy/pate.py]: add tensor_list type

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 7]
```diff
@@ -266,3 +278,245 @@ def perform_analysis(teacher_preds, indices, noise_eps, delta=1e-5, moments=8, b
...
+
+    """Converts list of torch tensors to list of integers/floats. Fix for not having the functionality which converts list of tensors to tensors
```
Comment (mari-linhares) [inline syft/frameworks/torch/differential_privacy/pate.py]: Change: *Converts list of torch tensors to list of integers/floats. Fix for not having the functionality which converts list of tensors to tensors*

to: *Converts a list of torch tensors to list of integers/floats.
Fix for not having the functionality which converts a list of tensors to tensors.*

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 8]
```diff
@@ -266,3 +278,245 @@ def perform_analysis(teacher_preds, indices, noise_eps, delta=1e-5, moments=8, b
...
+    
+       args:
+           
+           tensor_list[List]: List of torch tensors
```
Comment (mari-linhares) [inline syft/frameworks/torch/differential_privacy/pate.py]: tensor_list[List] -> tensor_list

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 9]
```diff
@@ -266,3 +278,245 @@ def perform_analysis(teacher_preds, indices, noise_eps, delta=1e-5, moments=8, b
...
+           
+       returns:
+           
+           literal_list[List]: List of floats/integers
```
Comment (mari-linhares) [inline syft/frameworks/torch/differential_privacy/pate.py]: literal_list[List] -> literal_list

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 10]
```diff
@@ -266,3 +278,245 @@ def perform_analysis(teacher_preds, indices, noise_eps, delta=1e-5, moments=8, b
...
+
+def tensors_to_literals(tensor_list):
+
+    """Converts list of torch tensors to list of integers/floats. Fix for not having the functionality which converts list of tensors to tensors
+    
+       args:
+           
+           tensor_list[List]: List of torch tensors
+           
+       returns:
```
Comment (mari-linhares) [inline syft/frameworks/torch/differential_privacy/pate.py]: returns -> Returns

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 11]
```diff
@@ -266,3 +278,245 @@ def perform_analysis(teacher_preds, indices, noise_eps, delta=1e-5, moments=8, b
...
+           
+           literal_list[List]: List of floats/integers
+           
+    """
+
+    literal_list = []
+
+    for tensor in tensor_list:
+
```
Comment (mari-linhares) [inline syft/frameworks/torch/differential_privacy/pate.py]: same about extra empty lines

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 12]
```diff
@@ -266,3 +278,245 @@ def perform_analysis(teacher_preds, indices, noise_eps, delta=1e-5, moments=8, b
...
+        literal_list.append(tensor.item())
+
+    return literal_list
+
+
+def logmgf_exact_torch(q, priv_eps, l):
+
```
Comment (mari-linhares) [inline syft/frameworks/torch/differential_privacy/pate.py]: remove empty line

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 13]
```diff
@@ -266,3 +278,245 @@ def perform_analysis(teacher_preds, indices, noise_eps, delta=1e-5, moments=8, b
...
+    """Computes the logmgf value given q and privacy eps.
+        The bound used is the min of three terms. The first term is from
```
Comment (mari-linhares) [inline syft/frameworks/torch/differential_privacy/pate.py]: Remove 1 indentation

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 14]
```diff
@@ -266,3 +278,245 @@ def perform_analysis(teacher_preds, indices, noise_eps, delta=1e-5, moments=8, b
...
+    else:
+
+        log_t = priv_eps * l
+
+    return min(0.5 * priv_eps * priv_eps * l * (l + 1), log_t, priv_eps * l)
+
+
+def compute_q_noisy_max_torch(counts, noise_eps):
+
+    """returns ~ Pr[outcome != winner].
```
Comment (mari-linhares) [inline syft/frameworks/torch/differential_privacy/pate.py]: returns -> Returns

File: test/torch/differential_privacy/test_pate.py [hunk 15]
```diff
@@ -15,3 +21,19 @@ def test_base_dataset():
     )
 
     assert data_dep_eps < data_ind_eps
+
+
+def test_base_dataset_torch():
```
[the remark this answers is not in the document]
Comment (kamathhrishi) [reply test/torch/differential_privacy/test_pate.py]: @iamtrask Done and tests are passing now :)

**Pull request merged**